src/exif.c

```diff
@@ -0,0 +1,124 @@
+// Copyright 2022 Google LLC. All rights reserved.
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "avif/internal.h"
+
+#include <stdint.h>
+#include <string.h>
+
+avifResult avifGetExifTiffHeaderOffset(const avifRWData * exif, uint32_t * offset)
+{
+    const uint8_t tiffHeaderBE[4] = { 'M', 'M', 0, 42 };
+    const uint8_t tiffHeaderLE[4] = { 'I', 'I', 42, 0 };
+    for (*offset = 0; *offset + 4 < exif->size; ++*offset) {
+        if (!memcmp(&exif->data[*offset], tiffHeaderBE, 4) || !memcmp(&exif->data[*offset], tiffHeaderLE, 4)) {
+            return AVIF_RESULT_OK;
+        }
+    }
+    // Couldn't find the TIFF header
+    return AVIF_RESULT_INVALID_EXIF_PAYLOAD;
+}
+
+avifResult avifImageExtractExifOrientationToIrotImir(avifImage * image)
+{
+    const avifTransformFlags otherFlags = image->transformFlags & ~(AVIF_TRANSFORM_IROT | AVIF_TRANSFORM_IMIR);
+    uint32_t offset;
+    const avifResult result = avifGetExifTiffHeaderOffset(&image->exif, &offset);
+    if (result != AVIF_RESULT_OK) {
+        // Couldn't find the TIFF header
+        return result;
+    }
+
+    avifROData raw = { image->exif.data + offset, image->exif.size - offset };
+    const avifBool littleEndian = (raw.data[0] == 'I');
+    avifROStream stream;
+    avifROStreamStart(&stream, &raw, NULL, NULL);
+
+    // TIFF Header
+    uint32_t offsetToNextIfd;
+    if (!avifROStreamSkip(&stream, 4) ||                                           // Skip tiffHeaderBE or tiffHeaderLE.
+        !avifROStreamReadU32Endianness(&stream, &offsetToNextIfd, littleEndian) || // Offset to 0th IFD
+        offsetToNextIfd < 4 + 4) {
+        return AVIF_RESULT_INVALID_EXIF_PAYLOAD;
+    }
+
+    while (offsetToNextIfd) { // for each IFD
```

Comment:
It seems that the while loop is not necessary, because orientation is in the 0th IFD.

Reply:
BUG: I just realized that if the first `offsetToNextIfd` is 8 and the second `offsetToNextIfd` is also 8, this while loop will be an infinite loop. We need to either remove the while loop (i.e., process only the 0th IFD) or defend against an infinite loop.

It would be good to create a test JPEG file with this condition.